src/range.cr

```diff
@@ -305,4 +305,46 @@ struct Range(B, E)
       self
     end
   end
+
+  #Perform binary search
+  #Based on Marc-André Lafortune's Ruby backports implementation, ported by T. Yamada
+  def bsearch
+    from = self.begin
+    to   = self.end
+    unless from.is_a?(Int) && to.is_a?(Int)
+      # Float support is currently dropped
+      raise "can't do binary search for #{from.class}"
+    end
+
+    midpoint = 0 # placeholder
+    convert = ->{ midpoint }
+
+    to -= 1 if excludes_end?
+    satisfied = nil
+    while from <= to
+      midpoint = (from + to)/2
+      result = yield(cur = convert.call)
```

Comment:
This is the most difficult point. Please take a look at https://github.com/marcandre/backports/blob/master/lib/backports/2.0.0/range/bsearch.rb#L19 . (As stated above) I tried `(pointerof(midpoint) as Pointer(Int64)).value`, but it caused internal error.